create task message using additional v2 parameters

jamesmallen opened the pull request:
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description
The `argsrepr` and `kwargsrepr` options described in the documentation (http://docs.celeryproject.org/en/latest/userguide/tasks.html#hiding-sensitive-information-in-arguments) are not actually used when creating task messages. This PR addresses that by checking the protocol version and passing them if using a version other than v1.

Fixes #4259

georgepsarakis commented on celery/app/base.py:
```diff
@@ -731,15 +731,26 @@ def send_task(self, name, args=None, kwargs=None, countdown=None,
-            maybe_list(link), maybe_list(link_error),
-            reply_to or self.oid, time_limit, soft_time_limit,
-            self.conf.task_send_sent_event,
-            root_id, parent_id, shadow, chain,
```
`AMQP.create_task_message` attempts to be protocol-agnostic, so I am not sure that we should apply a conditional based on protocol version. I would suggest to add an additional variadic keyworded-argument list in the signature of [AMQP.as_task_v1](https://github.com/celery/celery/blob/master/celery/app/amqp.py#L401):

```
def as_task_v1(..., **compat_kwargs):
```

Then both signatures will be able to receive the keyworded arguments `kwargsrepr` & `argsrepr`, but only `as_task_v2` will be actually using them.

What do you think?

jamesmallen replied:
Excellent point - for some reason I thought the `as_task` methods were coming from another library. I agree that it makes more sense to update those and remove the conditional. Updating now...

georgepsarakis commented on celery/app/amqp.py:
```diff
@@ -309,7 +309,8 @@ def as_task_v2(self, task_id, name, args=None, kwargs=None,
                    shadow=None, chain=None, now=None, timezone=None,
-                   origin=None, argsrepr=None, kwargsrepr=None):
+                   origin=None, argsrepr=None, kwargsrepr=None,
+                   **compat_kwargs):
```
Thanks for applying the suggested changes. Although it would work, you don't need `**compat_kwargs` in this signature, so I would recommend to perform as little changes as possible.

jamesmallen replied:
Sure thing - PR updated!

georgepsarakis commented:
@jamesmallen I am testing your branch locally and I do not see it fixing the issue. Do you have an example task and `apply_async` call that I can try? Thanks.

jamesmallen commented:
@georgepsarakis Here's an example task:
```
@app.task
def kwargsrepr_example(x=None):
    if x is None:
        kwargsrepr_example.apply_async(kwargs={'x': 'supersecret'}, kwargsrepr="{'x': '*******'}")
    print('x: %s', x)
```

Run this with `celery worker --loglevel DEBUG`.

Here's the output pre-patch:
```
[2017-09-18 15:58:53,053: INFO/MainProcess] Received task: tasks.kwargsrepr_example[5010f57d-4538-41f7-af45-744e4bb69ffd]
[2017-09-18 15:58:53,053: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x10de71268> (args:('tasks.kwargsrepr_example', '5010f57d-4538-41f7-af45-744e4bb69ffd', {'lang': 'py', 'task': 'tasks.kwargsrepr_example', 'id': '5010f57d-4538-41f7-af45-744e4bb69ffd', 'eta': None, 'expires': None, 'group': None, 'retries': 0, 'timelimit': [None, None], 'root_id': '5010f57d-4538-41f7-af45-744e4bb69ffd', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen82958@ip-10-10-200-58.us-east-2.compute.internal', 'reply_to': '6fe7b4f0-9fe1-3456-83b3-e3b41fe4ae06', 'correlation_id': '5010f57d-4538-41f7-af45-744e4bb69ffd', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": null, "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2017-09-18 15:58:53,055: DEBUG/MainProcess] Task accepted: tasks.kwargsrepr_example[5010f57d-4538-41f7-af45-744e4bb69ffd] pid:82624
[2017-09-18 15:58:53,074: DEBUG/PoolWorker-1] Start from server, version: 0.9, properties: {'capabilities': {'publisher_confirms': True, 'exchange_exchange_bindings': True, 'basic.nack': True, 'consumer_cancel_notify': True, 'connection.blocked': True, 'consumer_priorities': True, 'authentication_failure_close': True, 'per_consumer_qos': True, 'direct_reply_to': True}, 'cluster_name': 'rabbit@rabbit', 'copyright': 'Copyright (C) 2007-2017 Pivotal Software, Inc.', 'information': 'Licensed under the MPL.  See http://www.rabbitmq.com/', 'platform': 'Erlang/OTP', 'product': 'RabbitMQ', 'version': '3.6.10'}, mechanisms: ['PLAIN', 'AMQPLAIN'], locales: ['en_US']
[2017-09-18 15:58:53,076: DEBUG/PoolWorker-1] using channel_id: 1
[2017-09-18 15:58:53,079: DEBUG/PoolWorker-1] Channel open
[2017-09-18 15:58:53,081: WARNING/PoolWorker-1] x: %s
[2017-09-18 15:58:53,082: WARNING/PoolWorker-1] None
[2017-09-18 15:58:53,082: INFO/PoolWorker-1] Task tasks.kwargsrepr_example[5010f57d-4538-41f7-af45-744e4bb69ffd] succeeded in 0.02749929198762402s: None
[2017-09-18 15:58:53,083: INFO/MainProcess] Received task: tasks.kwargsrepr_example[f6a128f7-c947-4059-ab93-15d000d52080]
[2017-09-18 15:58:53,083: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x10de71268> (args:('tasks.kwargsrepr_example', 'f6a128f7-c947-4059-ab93-15d000d52080', {'lang': 'py', 'task': 'tasks.kwargsrepr_example', 'id': 'f6a128f7-c947-4059-ab93-15d000d52080', 'eta': None, 'expires': None, 'group': None, 'retries': 0, 'timelimit': [None, None], 'root_id': '5010f57d-4538-41f7-af45-744e4bb69ffd', 'parent_id': '5010f57d-4538-41f7-af45-744e4bb69ffd', 'argsrepr': '()', 'kwargsrepr': "{'x': 'supersecret'}", 'origin': 'gen82624@ip-10-10-200-58.us-east-2.compute.internal', 'reply_to': '172c1fec-7f43-3609-87e4-7493156352fe', 'correlation_id': 'f6a128f7-c947-4059-ab93-15d000d52080', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}}, '[[], {"x": "supersecret"}, {"callbacks": null, "errbacks": null, "chain": null, "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2017-09-18 15:58:53,084: DEBUG/MainProcess] Task accepted: tasks.kwargsrepr_example[f6a128f7-c947-4059-ab93-15d000d52080] pid:82626
[2017-09-18 15:58:53,085: WARNING/PoolWorker-3] x: %s
[2017-09-18 15:58:53,086: WARNING/PoolWorker-3] supersecret
[2017-09-18 15:58:53,087: INFO/PoolWorker-3] Task tasks.kwargsrepr_example[f6a128f7-c947-4059-ab93-15d000d52080] succeeded in 0.0025708680041134357s: None
```

Here's the output with the patch:
```
[2017-09-18 16:02:57,625: INFO/MainProcess] Received task: tasks.kwargsrepr_example[09841abd-5790-4173-a00a-3554e328703a]
[2017-09-18 16:02:57,626: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x1091cebf8> (args:('tasks.kwargsrepr_example', '09841abd-5790-4173-a00a-3554e328703a', {'lang': 'py', 'task': 'tasks.kwargsrepr_example', 'id': '09841abd-5790-4173-a00a-3554e328703a', 'eta': None, 'expires': None, 'group': None, 'retries': 0, 'timelimit': [None, None], 'root_id': '09841abd-5790-4173-a00a-3554e328703a', 'parent_id': None, 'argsrepr': '()', 'kwargsrepr': '{}', 'origin': 'gen84143@ip-10-10-200-58.us-east-2.compute.internal', 'reply_to': 'baf7f1d5-f81e-3b08-a32d-d4cf37cdcd60', 'correlation_id': '09841abd-5790-4173-a00a-3554e328703a', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}}, '[[], {}, {"callbacks": null, "errbacks": null, "chain": null, "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2017-09-18 16:02:57,628: DEBUG/MainProcess] Task accepted: tasks.kwargsrepr_example[09841abd-5790-4173-a00a-3554e328703a] pid:83972
[2017-09-18 16:02:57,649: DEBUG/PoolWorker-1] Start from server, version: 0.9, properties: {'capabilities': {'publisher_confirms': True, 'exchange_exchange_bindings': True, 'basic.nack': True, 'consumer_cancel_notify': True, 'connection.blocked': True, 'consumer_priorities': True, 'authentication_failure_close': True, 'per_consumer_qos': True, 'direct_reply_to': True}, 'cluster_name': 'rabbit@rabbit', 'copyright': 'Copyright (C) 2007-2017 Pivotal Software, Inc.', 'information': 'Licensed under the MPL.  See http://www.rabbitmq.com/', 'platform': 'Erlang/OTP', 'product': 'RabbitMQ', 'version': '3.6.10'}, mechanisms: ['PLAIN', 'AMQPLAIN'], locales: ['en_US']
[2017-09-18 16:02:57,652: DEBUG/PoolWorker-1] using channel_id: 1
[2017-09-18 16:02:57,656: DEBUG/PoolWorker-1] Channel open
[2017-09-18 16:02:57,669: WARNING/PoolWorker-1] x: %s
[2017-09-18 16:02:57,670: WARNING/PoolWorker-1] None
[2017-09-18 16:02:57,671: INFO/PoolWorker-1] Task tasks.kwargsrepr_example[09841abd-5790-4173-a00a-3554e328703a] succeeded in 0.04251735701109283s: None
[2017-09-18 16:02:57,671: INFO/MainProcess] Received task: tasks.kwargsrepr_example[b73d9913-a3b4-456d-baad-7e6e32631dc4]
[2017-09-18 16:02:57,671: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x1091cebf8> (args:('tasks.kwargsrepr_example', 'b73d9913-a3b4-456d-baad-7e6e32631dc4', {'lang': 'py', 'task': 'tasks.kwargsrepr_example', 'id': 'b73d9913-a3b4-456d-baad-7e6e32631dc4', 'eta': None, 'expires': None, 'group': None, 'retries': 0, 'timelimit': [None, None], 'root_id': '09841abd-5790-4173-a00a-3554e328703a', 'parent_id': '09841abd-5790-4173-a00a-3554e328703a', 'argsrepr': '()', 'kwargsrepr': "{'x': '*******'}", 'origin': 'gen83972@ip-10-10-200-58.us-east-2.compute.internal', 'reply_to': '60e1be15-50db-371f-b538-55b86a11aa42', 'correlation_id': 'b73d9913-a3b4-456d-baad-7e6e32631dc4', 'delivery_info': {'exchange': '', 'routing_key': 'celery', 'priority': 0, 'redelivered': False}}, '[[], {"x": "supersecret"}, {"callbacks": null, "errbacks": null, "chain": null, "chord": null}]', 'application/json', 'utf-8') kwargs:{})
[2017-09-18 16:02:57,673: DEBUG/MainProcess] Task accepted: tasks.kwargsrepr_example[b73d9913-a3b4-456d-baad-7e6e32631dc4] pid:83974
[2017-09-18 16:02:57,674: WARNING/PoolWorker-3] x: %s
[2017-09-18 16:02:57,675: WARNING/PoolWorker-3] supersecret
[2017-09-18 16:02:57,676: INFO/PoolWorker-3] Task tasks.kwargsrepr_example[b73d9913-a3b4-456d-baad-7e6e32631dc4] succeeded in 0.0021577919833362103s: None
```

georgepsarakis commented:
@jamesmallen thanks for providing the samples. I will try to think about a test case to verify this. Any feedback is appreciated of course :smile: !

jamesmallen commented:
@georgepsarakis I looked through the unit tests, and it appears that the task-focused unit tests all use task protocol v1. If that is in fact the case, now seems like a great time to get some tests put together for v2 I guess!

jamesmallen commented:
@georgepsarakis I've updated the tasks to use v2 of the task protocol, and was thus able to add tests for kwargsrepr/argsrepr (and also added in a test to confirm the default behavior when these aren't passed to `send_task`).

georgepsarakis commented:
@jamesmallen great, thank you!

jamesmallen commented:
@thedrow flake8 error addressed!
